python/ql/src/semmle/python/frameworks/MarkupSafe.qll

```diff
@@ -0,0 +1,135 @@
+/**
+ * Provides classes modeling security-relevant aspects of the `MarkupSafe` PyPI package.
+ * See https://markupsafe.palletsprojects.com/en/2.0.x/.
+ */
+
+private import python
+private import semmle.python.dataflow.new.DataFlow
+private import semmle.python.dataflow.new.TaintTracking
+private import semmle.python.Concepts
+private import semmle.python.ApiGraphs
+
+/**
+ * Provides models for the `MarkupSafe` PyPI package.
+ * See https://markupsafe.palletsprojects.com/en/2.0.x/.
+ */
+private module MarkupSafeModel {
+  /**
+   * Provides models for the `markupsafe.Markup` class
+   *
+   * See https://markupsafe.palletsprojects.com/en/2.0.x/escaping/#markupsafe.Markup.
+   */
+  module Markup {
+    /** Gets a reference to the `markupsafe.Markup` class. */
+    API::Node classRef() {
+      result = API::moduleImport("markupsafe").getMember("Markup")
+      or
+      result = API::moduleImport("flask").getMember("Markup")
+    }
+
+    /**
+     * A source of instances of `markupsafe.Markup`, extend this class to model new instances.
+     *
+     * This can include instantiations of the class, return values from function
+     * calls, or a special parameter that will be set when functions are called by an external
+     * library.
+     *
+     * Use the predicate `Markup::instance()` to get references to instances of `markupsafe.Markup`.
+     */
+    abstract class InstanceSource extends DataFlow::LocalSourceNode { }
+
+    /** A direct instantiation of `markupsafe.Markup`. */
+    private class ClassInstantiation extends InstanceSource, DataFlow::CallCfgNode {
+      override CallNode node;
+
+      ClassInstantiation() { this = classRef().getACall() }
+    }
+
+    /** Gets a reference to an instance of `markupsafe.Markup`. */
+    private DataFlow::LocalSourceNode instance(DataFlow::TypeTracker t) {
+      t.start() and
+      result instanceof InstanceSource
+      or
+      exists(DataFlow::TypeTracker t2 | result = instance(t2).track(t2, t))
+    }
+
+    /** Gets a reference to an instance of `markupsafe.Markup`. */
+    DataFlow::Node instance() { instance(DataFlow::TypeTracker::end()).flowsTo(result) }
+
+    /** A string concatenation with a `markupsafe.Markup` involved. */
+    class StringConcat extends Markup::InstanceSource, DataFlow::CfgNode {
+      override BinaryExprNode node;
+
+      StringConcat() {
+        node.getOp() instanceof Add and
+        instance().asCfgNode() in [node.getLeft(), node.getRight()]
+      }
+    }
+
+    /** A string format with `markupsafe.Markup` as the format string. */
+    class StringFormat extends Markup::InstanceSource, DataFlow::CallCfgNode {
+      StringFormat() {
+        exists(DataFlow::AttrRead attr | this.getFunction() = attr |
+          attr.getAttributeName() = "format" and
```

Comment:
Do we plan on supporting `%`-style formatting?

Also, would this be a place to use `MethodCallNode?

Reply:
`MethodCallNode` for sure :+1: I didn't initially plan on support `%`-style formatting, but we can do so :)

Reply:
https://github.com/github/codeql/pull/6092/commits/c2708176b140e56cfbcc752db03ca59ef45f24e8 adds %-style formatting :+1: